honeybee_radiance/writer.py

```diff
@@ -127,61 +137,356 @@ def room_to_rad(room, blk=False, minimal=False):
+        minimal: Boolean noting whether radiance strings should be written minimally.
+    """
+    if len(geometry) != 0 or len(geometry_blk) != 0:
+        # write the strings for the geometry
+        face_strs = []
+        if punched_verts:
+            for face in geometry:
+                modifier = face.properties.radiance.modifier
+                rad_poly = Polygon(face.identifier, face.punched_vertices, modifier)
+                face_strs.append(rad_poly.to_radiance(minimal, False, False))
+            for face, mod_name in zip(geometry_blk, mod_names):
+                modifier = mod_combs[mod_name][0]
+                rad_poly = Polygon(face.identifier, face.punched_vertices, modifier)
+                face_strs.append(rad_poly.to_radiance(minimal, False, False))
+        else:
+            for face in geometry:
+                modifier = face.properties.radiance.modifier
+                rad_poly = Polygon(face.identifier, face.vertices, modifier)
+                face_strs.append(rad_poly.to_radiance(minimal, False, False))
+            for face, mod_name in zip(geometry_blk, mod_names):
+                modifier = mod_combs[mod_name][0]
+                rad_poly = Polygon(face.identifier, face.vertices, modifier)
+                face_strs.append(rad_poly.to_radiance(minimal, False, False))
+
+        # write the strings for the modifiers
+        mod_strs = []
+        mod_blk_strs = []
+        for mod in modifiers:
+            if isinstance(mod, BSDF):
+                _process_bsdf_modifier(folder, mod, mod_strs, minimal)
+            else:
+                mod_strs.append(mod.to_radiance(minimal))
+        for mod in modifiers_blk:
+            if isinstance(mod, BSDF):
+                _process_bsdf_modifier(folder, mod, mod_strs, minimal)
+            else:
+                mod_blk_strs.append(mod.to_radiance(minimal))
+
+        # write the three files for the model sub-folder
+        dest = os.path.join(folder, sub_folder)
+        write_to_file_by_name(dest, '{}.rad'.format(model_id), '\n\n'.join(face_strs))
+        write_to_file_by_name(dest, '{}.mat'.format(model_id), '\n\n'.join(mod_strs))
+        write_to_file_by_name(dest, '{}.blk'.format(model_id), '\n\n'.join(mod_blk_strs))
+
+
+def _unique_modifiers(geometry_objects):
+    """Get a list of unique modifiers across an array of geometry objects.
+    
+    Args:
+        geometry_objects: An array of geometry objects (Faces, Apertures,
+            Doors, Shades) for which unique modifiers will be determined.
+
+    Returns:
+        A list of all unique modifiers across the input geometry_objects
+    """
+    modifiers = []
+    for obj in geometry_objects:
+        mod = obj.properties.radiance.modifier
+        if not _instance_in_array(mod, modifiers):
+            modifiers.append(mod)
+    return list(set(modifiers))
+
+
+def _unique_modifier_blk_combinations(geometry_objects):
```

Comment:
Interesting! why do we need the combinations here? Can't we just look for unique blacked and normal modifiers?

Reply:
Here's the situation:
I have two opaque objects with the same `modifier` but, on one of them, I have overridden the `modifier_blk`. Technically, there's only one modifier that "needs" to be written into the .mat file but I can't do this because there are two different modifiers to be written into the .blk file (one completely black modifier and the, the other, the overridden `modifer_blk`). So, I can't write just one modifier in tho the .mat file because the .mat and .blk files need to contain the same number of modifiers and the names of the modifiers need to match between the .mat and the .blk file.

So I need to duplicate the original modifier, change it's identifier to match that of the overridden `modifier_blk` and write two "duplicate" modifiers into the .mat file. This type of duplication only needs to happen every time there's a unique combination of `modifier` and overridden `modifier_blk` (otherwise, I can just use the same `modifier` and `modifier_blk` pair that already exists).

I hope that makes it clear.